File: packages/csharp/src/SheetCompressor/Compress.cs
using System.Collections.Generic;
using SheetCompressor.Encodings;

namespace SheetCompressor;

public static class Compressor
{
    public static CompressResult Compress(Grid grid, CompressOptions? options = null)
    {
        options ??= new CompressOptions();
        var strategy = Strategies.Resolve(options.AnchorStrategy);
        var detection = strategy.Detect(grid);
        var tokenCounter = options.TokenCounter ?? Tokens.EstimateTokens;
        var chartBlock = ChartDescriptors.RenderChartBlock(grid.Charts);

        var anchor = Encodings.Anchor.Encode(grid, detection, tokenCounter);
        var inverted = Encodings.InvertedIndex.Encode(grid, tokenCounter);
        var formatAgg = Encodings.FormatAggregation.Encode(grid, tokenCounter);

        return new CompressResult
        {
            Encodings = new EncodingsBundle
            {
                Anchor = WithCharts(anchor, chartBlock, tokenCounter),
                InvertedIndex = WithCharts(inverted, chartBlock, tokenCounter),
                FormatAggregation = WithCharts(formatAgg, chartBlock, tokenCounter),
            },
            Charts = grid.Charts is null
                ? new List<ChartDescriptor>()
                : new List<ChartDescriptor>(grid.Charts),
            RawBaseline = new TokenBaseline
            {
                TokenEstimate = tokenCounter(Baseline.VanillaEncode(grid)),
            },
        };
    }

    /// <summary>
    /// SPEC §6.2: extend an encoding's <c>.String</c> with the chart block (if
    /// any) and re-measure <c>.TokenEstimate</c> over the extended form. The
    /// <c>.Json</c> is unchanged — chart data only lives in the string + the
    /// top-level <c>Charts</c> echo.
    /// </summary>
    private static Encoding WithCharts(Encoding encoding, string chartBlock, TokenCounter tokenCounter)
    {
        if (chartBlock.Length == 0) return encoding;
        var s = ChartDescriptors.AppendChartBlock(encoding.String, chartBlock);
        return new Encoding
        {
            String = s,
            Json = encoding.Json,
            TokenEstimate = tokenCounter(s),
        };
    }
}

File: packages/go/sheetcompressor/compress.go
package sheetcompressor

// Compress is SPEC §2's top-level entry point. It runs all three encodings
// and returns the result struct verbatim. Both fields of `Options` are
// optional: nil AnchorStrategy → phase1; nil TokenCounter → the SPEC §7
// heuristic.
func Compress(grid *Grid, opts Options) CompressResult {
	strategy := ResolveStrategy(opts.AnchorStrategy)
	detection := strategy.Detect(grid)
	tokenCounter := opts.TokenCounter
	if tokenCounter == nil {
		tokenCounter = EstimateTokens
	}

	chartBlock := renderChartBlock(grid.Charts)

	anchor := encodeAnchor(grid, detection, tokenCounter)
	invIdx := encodeInvertedIndex(grid, tokenCounter)
	formatAgg := encodeFormatAggregation(grid, tokenCounter)

	anchor = withCharts(anchor, chartBlock, tokenCounter)
	invIdx = withChartsInv(invIdx, chartBlock, tokenCounter)
	formatAgg = withChartsFmt(formatAgg, chartBlock, tokenCounter)

	// `Charts` always materialises as a non-nil slice so JSON serialisation
	// emits `[]` rather than `null` when the input had no charts.
	charts := []ChartDescriptor{}
	if len(grid.Charts) > 0 {
		charts = append(charts, grid.Charts...)
	}

	res := CompressResult{
		Anchor:            anchor,
		InvertedIndex:     invIdx,
		FormatAggregation: formatAgg,
		Charts:            charts,
	}
	res.RawBaseline.TokenEstimate = tokenCounter(vanillaEncode(grid))
	return res
}

// withCharts / withChartsInv / withChartsFmt: SPEC §6.2 — splice the chart
// block onto the cell-string and re-measure tokens over the extended form.
// Three near-identical helpers because Go generics don't let us share a
// single function across the encoding-specific JSON types here.
func withCharts(e Encoding[AnchorJSON], block string, tc TokenCounter) Encoding[AnchorJSON] {
	if block == "" {
		return e
	}
	s := appendChartBlock(e.String, block)
	return Encoding[AnchorJSON]{String: s, JSON: e.JSON, TokenEstimate: tc(s)}
}

func withChartsInv(e Encoding[InvertedIndexJSON], block string, tc TokenCounter) Encoding[InvertedIndexJSON] {
	if block == "" {
		return e
	}
	s := appendChartBlock(e.String, block)
	return Encoding[InvertedIndexJSON]{String: s, JSON: e.JSON, TokenEstimate: tc(s)}
}

func withChartsFmt(e Encoding[FormatAggregationJSON], block string, tc TokenCounter) Encoding[FormatAggregationJSON] {
	if block == "" {
		return e
	}
	s := appendChartBlock(e.String, block)
	return Encoding[FormatAggregationJSON]{String: s, JSON: e.JSON, TokenEstimate: tc(s)}
}
